server/src/main/java/org/elasticsearch/index/mapper/DocumentParserListener.java

```diff
@@ -0,0 +1,192 @@
+                return parser.textOffset();
+            }
+        }
+
+        non-sealed interface ValueToken<T> extends Token {
+            T value() throws IOException;
+        }
+
+        Token START_OBJECT = new StartObject();
+        Token END_OBJECT = new EndObject();
+        Token START_ARRAY = new StartArray();
+        Token END_ARRAY = new EndArray();
+
+        static Token current(XContentParser parser) throws IOException {
+            return switch (parser.currentToken()) {
+                case START_OBJECT -> Token.START_OBJECT;
+                case END_OBJECT -> Token.END_OBJECT;
+                case START_ARRAY -> Token.START_ARRAY;
+                case END_ARRAY -> Token.END_ARRAY;
+                case FIELD_NAME -> new FieldName(parser.currentName());
+                case VALUE_STRING -> {
+                    if (parser.hasTextCharacters()) {
+                        yield new StringAsCharArrayValue(parser);
+                    } else {
+                        yield (ValueToken<String>) parser::text;
+                    }
+                }
+                case VALUE_NUMBER -> switch (parser.numberType()) {
+                    case INT -> (ValueToken<Integer>) parser::intValue;
+                    case BIG_INTEGER -> (ValueToken<BigInteger>) () -> (BigInteger) parser.numberValue();
+                    case LONG -> (ValueToken<Long>) parser::longValue;
+                    case FLOAT -> (ValueToken<Float>) parser::floatValue;
+                    case DOUBLE -> (ValueToken<Double>) parser::doubleValue;
+                    case BIG_DECIMAL -> {
+                        // See @XContentGenerator#copyCurrentEvent
+                        assert false : "missing xcontent number handling for type [" + parser.numberType() + "]";
+                        yield null;
+                    }
+                };
+                case VALUE_BOOLEAN -> (ValueToken<Boolean>) parser::booleanValue;
+                case VALUE_EMBEDDED_OBJECT -> (ValueToken<byte[]>) parser::binaryValue;
+                case VALUE_NULL -> new NullValue();
+                case null -> null;
+            };
+        }
+    }
+
+    sealed interface Event permits Event.DocumentStart, Event.ObjectStart, Event.ObjectEnd, Event.ObjectArrayStart, Event.ObjectArrayEnd,
+        Event.LeafArrayStart, Event.LeafArrayEnd, Event.LeafValue {
+        record DocumentStart(RootObjectMapper rootObjectMapper, LuceneDocument document) implements Event {}
+
+        record ObjectStart(ObjectMapper objectMapper, boolean insideObjectArray, ObjectMapper parentMapper, LuceneDocument document)
+            implements
+                Event {}
+
+        record ObjectEnd(ObjectMapper objectMapper) implements Event {}
+
+        record ObjectArrayStart(ObjectMapper objectMapper, ObjectMapper parentMapper, LuceneDocument document) implements Event {}
+
+        record ObjectArrayEnd() implements Event {}
+
+        final class LeafValue implements Event {
```

Comment:
Nit: make this a record too?

Reply:
I don't want to expose `parser` from it.